blackjack.py

```diff
@@ -104,17 +113,16 @@ def get_hard_hand_value(self):
         :return: int
         """
         hard_hand_value = 0
-        for card in self.hand:
-            if self.hand[0].value == 'Ace' and self.hand[1].value in ['Jack', 'Queen', 'King']:
-                hard_hand_value = 21.1
-            elif self.hand[0].value in ['Jack', 'Queen', 'King'] and self.hand[1].value == 'Ace':
-                hard_hand_value = 21.1
-            elif card.value.isdigit():
-                hard_hand_value += int(card.value)
-            elif card.value == 'Ace':
-                hard_hand_value += 11
-            else:
-                hard_hand_value += 10
+        if self.check_natural_win():
+            hard_hand_value = 21.1
+        else:
+            for card in self.hand:
+                if card.value.isdigit():
+                    hard_hand_value += int(card.value)
+                elif card.value == 'Ace':
+                    hard_hand_value += 11
+                else:
+                    hard_hand_value += 10
```

Comment:
@romainrichard 
Thank you for reviewing, that's some great feedback! 🙂 
So when you say: "By leveraging the power of OOP a bit more you will make your life so much easier", can you explain a bit further how I can work on doing this or give me an example? 

I know it's sort of a difficult thing to put succinctly so if you think this is the best way to learn, I could write a small project of your choice that will use OOP and then we can use that to show how I could be utilising OOP better? 

I know I'm not getting the most out of OOP that I could be, just not sure where to start or what bad signs to look out for in my code

Reply:
Look into [dunder methods](https://dbader.org/blog/python-dunder-methods) (also named magic methods), those can really help make your code more readable.

For example, since we were talking about vectors in the chat, this would be without dunder methods:
```
vector1.add(vector2)
```
and this would be with dunder methods:
```
vector1 + vector2
```

In your case, one thing that could be improved is the use of `.value` everywhere.
Maybe instead of writing:
```
if self.hand[0].value == 'Ace':
```
you'd like to write:
```
if self.hand[0] == 'Ace':
```
It looks a bit better, but comparing a `Card` to a `string` doesn't feel right, so maybe we could do something like this:
```
ACE_CARD = Card('Ace', None)  # We don't care about the suit
if self.hand[0] == ACE_CARD:
```
So how would we make this work? We need to implement our own `==` operation on the `Card` class as this is the test we are doing, so we'll need to implement the `__eq__` dunder method.
That method takes 2 parameters: `(self, other)`. `self` is the first `Card` object and `other` is the one we're comparing it to. It returns a boolean, `True` if `self` and `other` should be considered equal and `False` otherwise.

Let's dive into it! First, let's see what we get with the current code:
```
In [1]: c1 = Card("Ace", "Hearts")

In [2]: c2 = Card("Ace", "Hearts")

In [3]: c1 == c2
Out[3]: False
```
So `c1` and `c2` seem like they should be equal, but they're not. Let's fix that:
```
class Card(object):
    def __eq__(self, other):
        return True
```
Now if you try the test again, `c1` and `c2` will be considered equal, which is good... But all `Card` instances are equal now because `__eq__` always returns `True`, so we need to define how we want to equality to actually work.
The most logical way of checking equality is to check if the `value` and `suit` of the two cards are the same.
```
class Card(object):
    def __eq__(self, other):
        return self.value == other.value and self.suit == other.suit
```
That makes sense and works fine, but it doesn't fit our example.
We want to be able to check if a card is an Ace, no matter the suit. So let's not check the `suit` attribute.
```
class Card(object):
    def __eq__(self, other):
        return self.value == other.value
```
Alright, now our example works and we could stop here.
But it may feels strange because now `Card('Ace', 'Hearts') == Card('Ace', 'Diamonds')` is `True`, so maybe there's a better way to do this.

Maybe we could say that we don't ignore the `suit` attribute if its value on both `self` and `other` is properly defined. As in, if they're not `None`, we should compare them. And if one of the card's `suit` attribute is `None`, then we'll only compare the `value`.
```
class Card(object):
    def __eq__(self, other):
        return self.value == other.value and (
            self.suit is None or other.suit is None or self.suit == other.suit
        )
```
Which is fine, but it can me more pythonic:
```
class Card(object):
    def __eq__(self, other):
        return self.value == other.value and (
            None in (self.suit, other.suit) or self.suit == other.suit
        )
```
We could do the same thing for the `value` attribute, i.e. allowing cards with a `value` of `None` and taking that into consideration for comparison, but there's no reason to implement it if we don't have a need for it ([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)).

So now, we can drop the `.value` everywhere in the code which will make your code cleaner.